src/Storages/FileLog/StorageFileLog.cpp

```diff
@@ -0,0 +1,432 @@
+
+        auto path = path_ast->as<ASTLiteral &>().value.safeGet<String>();
+        auto format = format_ast->as<ASTLiteral &>().value.safeGet<String>();
+
+        return StorageFileLog::create(args.table_id, args.getContext(), args.columns, path, format, std::move(filelog_settings));
+    };
+
+    factory.registerStorage(
+        "FileLog",
+        creator_fn,
+        StorageFactory::StorageFeatures{
+            .supports_settings = true,
+        });
+}
+
+bool StorageFileLog::updateFileStatus()
+{
+    /// Do not need to hold file_status lock, since it will be holded
+    /// by caller when call this function
+    auto error = directory_watch->hasError();
+    if (error)
+        LOG_INFO(log, "Error happened during watching directory {}.", directory_watch->getPath());
+
+    auto events = directory_watch->getEvents();
+
+    for (const auto & event : events)
+    {
+        switch (event.type)
+        {
+            case Poco::DirectoryWatcher::DW_ITEM_ADDED: {
+                auto normal_path = std::filesystem::path(event.path).lexically_normal().native();
+                LOG_TRACE(log, "New event {} watched, path: {}", event.callback, event.path);
+                if (std::filesystem::is_regular_file(normal_path))
+                {
+                    file_status[event.path] = FileContext{};
+                    file_names.push_back(normal_path);
+                }
+                break;
+            }
+
+            case Poco::DirectoryWatcher::DW_ITEM_MODIFIED: {
+                auto normal_path = std::filesystem::path(event.path).lexically_normal().native();
+                LOG_TRACE(log, "New event {} watched, path: {}", event.callback, normal_path);
+                if (std::filesystem::is_regular_file(normal_path) && file_status.contains(normal_path))
+                {
+                    file_status[normal_path].status = FileStatus::UPDATED;
+                }
+                break;
+            }
+
+            case Poco::DirectoryWatcher::DW_ITEM_REMOVED:
+            case Poco::DirectoryWatcher::DW_ITEM_MOVED_TO:
+            case Poco::DirectoryWatcher::DW_ITEM_MOVED_FROM: {
+                auto normal_path = std::filesystem::path(event.path).lexically_normal().native();
+                LOG_TRACE(log, "New event {} watched, path: {}", event.callback, normal_path);
+                if (std::filesystem::is_regular_file(normal_path) && file_status.contains(normal_path))
+                {
+                    file_status[normal_path].status = FileStatus::REMOVED;
+                }
+            }
+        }
+    }
+    std::vector<String> valid_files;
+    for (const auto & file_name : file_names)
+    {
+        if (file_status.at(file_name).status == FileStatus::REMOVED)
+        {
+            file_status.erase(file_name);
+        }
+        else
+        {
+            valid_files.push_back(file_name);
+        }
+    }
+
+    file_names.swap(valid_files);
+
+    return events.empty() || file_names.empty();
```

Comment:
Looks like we never use the return value, may be make it void then?

Reply:
`streamToViews` uses this to judge whether have file update or not.